### DISCUSSION TITLE
uint64_t with chipKIT

### DISCUSSION THREAD

**Discussion by JacobChrist** [chipKIT32/chipKIT-core#482] [Q&A]
I've seen some one off issues with uint64_t with the fubarino mini 2.0.  I think this is the first time ever that I've used a uint64_t anytime.  I have this function, and doesn't seem to always shift the correct amount of time.  I don't feel like I need the type casting on the 3 or that pos needs to be a uint64_t but I was grasping for straws.

```
uint64_t LED_R = 0x100;
uint64_t LED_G = 0x010;
uint64_t LED_B = 0x001;

    uint64_t dot_color(uint64_t color, uint64_t pos)
    {
      if (pos > 15) pos = 15;
      return ((uint64_t)color) << (pos * (uint64_t)3);
    }
```
This code is being called the following code that read a 3-axis accelerometer and then displays one color per axis on a 16 pixel RGB dispay.  X gets shifted two more times that pos_x, Y gets shifted one more time that pos_y and Z is correct.  It so strange.  I'm burnt out for the day, but when I get back to this I'm going to try using utint32_t to see if I have better results.
```
void accel_level(void)
{
  float Ax, Ax_min = -1, Ax_max = 1;
  float Ay, Ay_min = -1, Ay_max = 1;
  float Az, Az_min = -1, Az_max = 1;

  while (Serial.available() == 0 && digitalRead(PGM_BUTTON) != LOW)
  {
    if (accel.available())
    {
      Ax = accel.getX(); // range is about -1100.0 to 1100.0
      Ay = accel.getY();
      Az = accel.getZ();
      if (Ax > Ax_max) Ax_max = Ax;
      if (Ax < Ax_min) Ax_min = Ax;
      if (Ay > Ay_max) Ay_max = Ay;
      if (Ay < Ay_min) Ay_min = Ay;
      if (Az > Az_max) Az_max = Az;
      if (Az < Az_min) Az_min = Az;
    }
    uint8_t pos_x = (uint8_t)map(Ax, Ax_min, Ax_max, 15, 0);
    uint8_t pos_y = (uint8_t)map(Ay, Ay_min, Ay_max, 15, 0);
    uint8_t pos_z = (uint8_t)map(Az, Az_min, Az_max, 15, 0);
    uint64_t color = 0x0000;
    color |= led.dot_color( LED_R, pos_x );
    color |= led.dot_color( LED_G, pos_y );
    color |= led.dot_color( LED_B, pos_z );
    led.led_out( color );
  }

  while (digitalRead(PGM_BUTTON) == LOW);
  delay(100);
}
```
Jacob

**Comment by majenkotech** [comment]
I would guess it's related to your pos_* variables being only 8 bit and you're passing them through 64 bit parameters. Try passing that parameter as 8 bit not 64 bit (64 bit is wasteful for that parameter anyway).

There should be no need to do any casting at all for a simple bit shift.

You can also try looking at the generated assembly code to see what it thinks it's doing...
